src/contracts/rewardsV2/VaultSnapshotRewards.sol

```diff
@@ -0,0 +1,422 @@
+            revert InvalidHintsLength();
+        }
+
+        uint256 amount;
+        uint256 rewardIndex = startIndex;
+        for (uint256 i; i < rewardsToClaim; ++i) {
+            RewardDistribution storage reward = rewardsByTokenNetwork[rewardIndex];
+
+            amount += IVault(vault).activeSharesOfAt(msg.sender, reward.timestamp, hints[i])
+                .mulDiv(reward.amount, _vaultSnapshotRewardsStorage()._activeSharesCache[vault][reward.timestamp]);
+
+            ++rewardIndex;
+        }
+
+        _vaultSnapshotRewardsStorage()._lastUnclaimedReward[msg.sender][vault][network][token] = rewardIndex;
+
+        if (amount > 0) {
+            IERC20(token).safeTransfer(recipient, amount);
+        }
+
+        emit ClaimVaultSnapshotRewards(msg.sender, network, token, vault, amount, lastUnclaimedReward_);
+    }
+
+    /**
+     * @inheritdoc IVaultSnapshotRewards
+     */
+    function claimCuratorFee(
+        address recipient,
+        address vault,
+        address token
+    ) public {
+        address curator = ICuratorRegistry(CURATOR_REGISTRY).getCurator(vault);
+        if (curator != msg.sender) {
+            revert NotCurator();
+        }
+
+        uint256 claimableFee = _vaultSnapshotRewardsStorage()._curatorFees[vault][token];
+        if (claimableFee == 0) {
+            revert NoRewardsToClaim();
+        }
+
+        _vaultSnapshotRewardsStorage()._curatorFees[vault][token] = 0;
+        IERC20(token).safeTransfer(recipient, claimableFee);
+
+        emit ClaimCuratorFee(vault, token, claimableFee);
+    }
+
+    /**
+     * @inheritdoc IVaultSnapshotRewards
+     */
+    function claimOperatorFee(
+        address recipient,
+        address network,
+        address token,
+        address vault,
+        uint256 lastUnclaimedRewards,
+        uint256 firstRewardToClaim,
+        uint256 maxRewards,
+        bytes calldata extraData
+    ) public {
+        if (recipient == address(0)) {
+            revert InvalidRecipient();
+        }
+
+        // Check lastUnclaimedRewards vs on-chain for reorgs
+        uint256 lastUnclaimedReward_ =
+            _vaultSnapshotRewardsStorage()._lastUnclaimedOperatorReward[msg.sender][vault][network][token];
+        if (lastUnclaimedRewards != lastUnclaimedReward_) {
+            revert InvalidHintsLength(); // Reusing error
+        }
+
+        RewardDistribution[] storage rewardsByTokenNetwork =
+            _vaultSnapshotRewardsStorage()._rewards[vault][network][token];
+
+        uint256 startIndex = firstRewardToClaim > 0 ? firstRewardToClaim : lastUnclaimedReward_;
+        if (startIndex > rewardsByTokenNetwork.length) {
+            revert NoRewardsToClaim();
+        }
+
+        uint256 rewardsToClaim = Math.min(maxRewards, rewardsByTokenNetwork.length - startIndex);
+
+        if (rewardsToClaim == 0) {
+            revert NoRewardsToClaim();
+        }
+
+        address delegatorAddress = IVault(vault).delegator();
```

Comment:
Let's adjust distribution struct to:

struct RewardDistribution {
      uint96 subnetworkId;
			address delegator;
			uint64 delgatorType;
      uint48 timestamp;
      uint256 amount;
      uint256 operatorsFee;
  }
  
  It'll cover possible future func to change delegator address, and make the claiming cheaper